File: gin_chat/room_channel.go
package main

import (
	"encoding/json"
	"fmt"
	"log"

	"github.com/piecehealth/actioncable"
)

var getRoomId = func(c *actioncable.Channel) string {
	params := struct {
		Id string `json:"id"`
	}{}

	json.Unmarshal(c.Params, &params)

	return fmt.Sprintf("room_%s", params.Id)
}

var roomChannel = &actioncable.ChannelDescripion{
	Name: "RoomChannel",
	Subscribed: func(c *actioncable.Channel) {
		roomId := getRoomId(c)
		currentUser := fmt.Sprintf("%v", c.ConnIdentifier)

		if roomId == "forbidden_room" && currentUser != "the chosen one" {
			c.Reject()
			return
		}

		log.Printf("%v just join the %s.", c.ConnIdentifier, roomId)

		c.StreamFrom(roomId)

		m := map[string]string{
			"user_join": currentUser + " just join the room.",
		}
		c.Broadcast(roomId, m)
	},
	Unsubscribed: func(c *actioncable.Channel) {
		roomId := getRoomId(c)

		log.Printf("%v just left the %s.", c.ConnIdentifier, roomId)

		m := map[string]string{
			"user_leave": fmt.Sprintf("%v", c.ConnIdentifier) + " just left the room.",
		}
		cable.Broadcast("RoomChannel", roomId, m)
	},
	PerformAction: func(c *actioncable.Channel, data string) {
		d := struct {
			Action  string `json:"action"`
			Message string `json:"message"`
			Name    string `json:"name"`
		}{}

		json.Unmarshal([]byte(data), &d)

		roomId := getRoomId(c)

		switch d.Action {
		case "send_message":
			currentUser := fmt.Sprintf("%v", c.ConnIdentifier)

			m := map[string]string{
				"send_by": currentUser,
				"message": d.Message,
			}

			c.Broadcast(roomId, m)
		case "kick":
			log.Printf("disconnect %s's connection.", d.Name)
			cable.DisconnectRemoteConnection(d.Name)
		case "stop_stream":
			log.Println("stop all streams.")
			c.StopAllStreams()
		default:
			log.Printf("Unable to process RoomChannel#%s", d.Action)
		}
	},
}

File: rails_chat/app/channels/room_channel.rb
class RoomChannel < ApplicationCable::Channel
  def subscribed
    if params[:id] == "forbidden_room" && current_user != "the chosen one"
      reject

      return
    end

    stream_from room_id
    Rails.logger.info("#{current_user} just join the #{room_id}.")
    ActionCable.server.broadcast(room_id, {user_join: "#{current_user} just join the room."})
  end

  def unsubscribed
    Rails.logger.info("#{current_user} just left the #{room_id}.")
    ActionCable.server.broadcast(room_id, {user_leave:  "#{current_user} just left the room."})
  end

  def send_message(data)
    ActionCable.server.broadcast(room_id, {send_by: current_user, message: data["message"]})
  end

  def kick(data)
    Rails.logger.info("disconnect #{data["name"]}'s connection.")
    ActionCable.server.remote_connections.where(current_user: data["name"]).disconnect
  end

  def stop_stream
    Rails.logger.info("stop all streams.")
    stop_all_streams
  end

  private
    def room_id
      "room_#{params[:id]}"
    end
end
